vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/decode/FFmpegVideoDecoder.kt

```diff
@@ -0,0 +1,230 @@
+package org.vitrivr.engine.index.decode
+
+import com.github.kokorin.jaffree.StreamType
+import com.github.kokorin.jaffree.ffmpeg.*
+import com.github.kokorin.jaffree.ffprobe.FFprobe
+import io.github.oshai.kotlinlogging.KLogger
+import io.github.oshai.kotlinlogging.KotlinLogging
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel.Factory.RENDEZVOUS
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.channelFlow
+import org.bytedeco.javacv.FrameGrabber
+import org.vitrivr.engine.core.context.IndexContext
+import org.vitrivr.engine.core.model.relationship.Relationship
+import org.vitrivr.engine.core.model.retrievable.Ingested
+import org.vitrivr.engine.core.model.retrievable.Retrievable
+import org.vitrivr.engine.core.model.retrievable.attributes.ContentAuthorAttribute
+import org.vitrivr.engine.core.model.retrievable.attributes.SourceAttribute
+import org.vitrivr.engine.core.model.retrievable.attributes.time.TimeRangeAttribute
+import org.vitrivr.engine.core.operators.ingest.Decoder
+import org.vitrivr.engine.core.operators.ingest.DecoderFactory
+import org.vitrivr.engine.core.operators.ingest.Enumerator
+import org.vitrivr.engine.core.source.MediaType
+import org.vitrivr.engine.core.source.file.FileSource
+import java.awt.image.BufferedImage
+import java.nio.file.Path
+import java.util.*
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+import org.vitrivr.engine.core.source.Metadata
+
+class FFmpegVideoDecoder : DecoderFactory {
+
+    override fun newDecoder(name: String, input: Enumerator, context: IndexContext): Decoder {
+        val maxWidth = context[name, "maxWidth"]?.toIntOrNull() ?: 3840
+        val maxHeight = context[name, "maxHeight"]?.toIntOrNull() ?: 2160
+        val framerate = context[name, "framerate"]?.toIntOrNull()
```

Comment:
What's the idea of this parameter? And shouldn't the parameters for `VideoDecoder` and `FFmpegVideoDecoder` be aligned (more or less).

Reply:
Ideally, yes. The parameters I added here appeared to be useful and they are not currently supported by the other decoder. I also don't know if the library used in the other decoder supports those options. As for the alignment in the other direction, I'm frankly not entirely certain what the options do...

Reply:
I'm frankly not convinced that these parameters are something we want to offer at all. If I read them correctly, you can use them to downsample the video during decoding. Do we really need this?

I'd argue that one could come-up with a great deal of different filters that one could apply in this decoding step when using ffmpeg. Why only downsampling the resolution? Why not convert to b/w, for example? 

If anything, I'd add a parameter here that can be used to define invocation arguments for the ffmpeg process. Then people can do whatever they like without cluttering the parameter space.

Reply:
sure, more is always an option